x/iro/keeper/trade.go

```diff
@@ -0,0 +1,215 @@
+	costCoin := sdk.NewCoin(appparams.BaseDenom, cost)
+
+	totalCost, takerFeeCoin := k.SubtractTakerFee(costCoin, k.GetParams(ctx).TakerFee)
+	if !totalCost.IsPositive() || !takerFeeCoin.IsPositive() {
+		return errorsmod.Wrapf(types.ErrInvalidCost, "totalCost: %s, takerFeeCoin: %s", totalCost.String(), takerFeeCoin.String())
+	}
+
+	// Validate expected out amount
+	if totalCost.Amount.LT(minCost) {
+		return errorsmod.Wrapf(types.ErrInvalidMinCost, "minCost: %s, cost: %s", minCost.String(), totalCost.String())
+	}
+
+	// Charge taker fee
+	err = k.chargeTakerFee(ctx, takerFeeCoin, seller)
+	if err != nil {
+		return err
+	}
+
+	// send allocated tokens from seller to the plan
+	err = k.BK.SendCoinsFromAccountToModule(ctx, seller, types.ModuleName, sdk.NewCoins(sdk.NewCoin(plan.TotalAllocation.Denom, amountTokensToSell)))
+	if err != nil {
+		return err
+	}
+
+	// send DYM from the plan to the seller. DYM managed by the plan's module account
+	err = k.BK.SendCoins(ctx, plan.GetAddress(), seller, sdk.NewCoins(costCoin))
+	if err != nil {
+		return err
+	}
+
+	// Update plan
+	plan.SoldAmt = plan.SoldAmt.Sub(amountTokensToSell)
+	k.SetPlan(ctx, plan)
+
+	// Emit event
+	err = ctx.EventManager().EmitTypedEvent(&types.EventSell{
+		Seller:    seller.String(),
+		PlanId:    planId,
+		RollappId: plan.RollappId,
+		Amount:    amountTokensToSell,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) validateIROTradeable(ctx sdk.Context, plan types.Plan, trader string) error {
+	if plan.IsSettled() {
+		return errorsmod.Wrapf(types.ErrPlanSettled, "planId: %d", plan.Id)
+	}
+
+	// Validate start time started (unless the trader is the owner)
+	if ctx.BlockTime().Before(plan.StartTime) && k.rk.MustGetRollapp(ctx, plan.RollappId).Owner != trader {
+		return errorsmod.Wrapf(types.ErrPlanNotStarted, "planId: %d", plan.Id)
+	}
+
+	return nil
+}
+
+func (k Keeper) chargeTakerFee(ctx sdk.Context, takerFee sdk.Coin, sender sdk.AccAddress) error {
+	return k.BK.SendCoinsFromAccountToModule(ctx, sender, txfeestypes.ModuleName, sdk.NewCoins(takerFee))
+}
+
+// AddTakerFee returns the remaining amount after subtracting the taker fee and the taker fee amount
+// returns (1 + takerFee) * tokenIn, takerFee * tokenIn
+func (k Keeper) AddTakerFee(amt sdk.Coin, takerFee sdk.Dec) (sdk.Coin, sdk.Coin) {
+	takerFeeAmt := math.LegacyNewDecFromInt(amt.Amount).Mul(takerFee).TruncateInt()
+	newAmt := amt.Amount.Add(takerFeeAmt)
+	return sdk.NewCoin(amt.Denom, newAmt), sdk.NewCoin(amt.Denom, takerFeeAmt)
+}
+
+// SubtractTakerFee returns the remaining amount after subtracting the taker fee and the taker fee amount
+// returns (1 - takerFee) * tokenIn, takerFee * tokenIn
+func (k Keeper) SubtractTakerFee(amt sdk.Coin, takerFee sdk.Dec) (sdk.Coin, sdk.Coin) {
+	takerFeeAmt := math.LegacyNewDecFromInt(amt.Amount).Mul(takerFee).TruncateInt()
+	newAmt := amt.Amount.Sub(takerFeeAmt)
+	return sdk.NewCoin(amt.Denom, newAmt), sdk.NewCoin(amt.Denom, takerFeeAmt)
+}
```

Comment:
can be dried out:

func (k Keeper) ApplyTakerFee(amt sdk.Coin, takerFee sdk.Dec, isAdd bool) (sdk.Coin, sdk.Coin) {
	feeAmt := math.LegacyNewDecFromInt(amt.Amount).Mul(takerFee).TruncateInt()
	feeCoin := sdk.NewCoin(amt.Denom, feeAmt)
	
	var newAmt sdk.Int
	if isAdd {
		newAmt = amt.Amount.Add(feeAmt)
	} else {
		newAmt = amt.Amount.Sub(feeAmt)
	}
	
	return sdk.NewCoin(amt.Denom, newAmt), feeCoin
}